packages/server/graphql/types/User.ts

```diff
@@ -554,21 +554,41 @@ const User: GraphQLObjectType<any, GQLContext> = new GraphQLObjectType<any, GQLC
     },
     teams: {
       type: new GraphQLNonNull(new GraphQLList(new GraphQLNonNull(Team))),
-      description: 'all the teams the user is on that the viewer can see.',
+      description: 'all the teams visible to the user',
       resolve: async (
         {id: userId}: {id: string},
         _args: unknown,
         {authToken, dataLoader}: GQLContext
       ) => {
         const viewerId = getUserId(authToken)
-        const user = (await dataLoader.get('users').load(userId))!
-        const teamIds =
-          viewerId === userId || isSuperUser(authToken)
-            ? user.tms
-            : user.tms.filter((teamId: string) => authToken.tms.includes(teamId))
-        const teams = (await dataLoader.get('teams').loadMany(teamIds)).filter(isValid)
-        teams.sort((a, b) => (a.name > b.name ? 1 : -1))
-        return teams
+        const [user, organizationUsers] = await Promise.all([
+          dataLoader.get('users').load(userId),
+          dataLoader.get('organizationUsersByUserId').load(userId)
+        ])
+        const orgIds = organizationUsers.map(({orgId}) => orgId)
+
+        const organizationTmsWithPublicFlag = []
+
+        for (const orgId of orgIds) {
+          const organization = await dataLoader.get('organizations').load(orgId)
+          if (organization.featureFlags?.includes('publicTeams')) {
+            const tms = await dataLoader.get('teamsByOrgIds').load(organization.id)
+            organizationTmsWithPublicFlag.push(
+              ...tms.filter((team) => !user?.tms.includes(team.id))
+            )
+          }
+        }
+
+        const isViewerOrSuperUser = viewerId === userId || isSuperUser(authToken)
```

Comment:
+1 putting this into a separate variable is ok, but does not add much readability as it's only used once.